pkg/apiserver/middlewares/v1/api_key.go

```diff
@@ -49,34 +53,143 @@ func HashSHA512(str string) string {
 
 func (a *APIKey) MiddlewareFunc() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		val, ok := c.Request.Header[APIKeyHeader]
-		if !ok {
-			c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
-			c.Abort()
-			return
-		}
+		var bouncer *ent.Bouncer
+		var err error
 
-		hashStr := HashSHA512(val[0])
-		bouncer, err := a.DbClient.SelectBouncer(hashStr)
-		if err != nil {
-			log.Errorf("auth api key error: %s", err)
-			c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
-			c.Abort()
-			return
+		if c.Request.TLS != nil && len(c.Request.TLS.PeerCertificates) > 0 {
+			if a.TlsAuth == nil {
+				log.WithField("ip", c.ClientIP()).Error("TLS Auth is not configured but client presented a certificate")
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			validCert, extractedCN, err := a.TlsAuth.ValidateCert(c)
+			if !validCert {
+				log.WithField("ip", c.ClientIP()).Errorf("invalid client certificate: %s", err)
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			if err != nil {
+				log.WithField("ip", c.ClientIP()).Error(err)
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			bouncerName := fmt.Sprintf("%s@%s", extractedCN, c.ClientIP())
+			bouncer, err = a.DbClient.SelectBouncerByName(bouncerName)
+			//This is likely not the proper way, but isNotFound does not seem to work
+			if err != nil && strings.Contains(err.Error(), "bouncer not found") {
+				//Because we have a valid cert, automatically create the bouncer in the database if it does not exist
+				//Set a random API key, but it will never be used
+				//To fix
+				apiKey, err := GenerateAPIKey(64)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"ip": c.ClientIP(),
+						"cn": extractedCN,
+					}).Errorf("error generating mock api key: %s", err)
+					c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+					c.Abort()
+					return
+				}
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+					"cn": extractedCN,
+				}).Infof("Creating bouncer %s", bouncerName)
+				err = a.DbClient.CreateBouncer(bouncerName, c.ClientIP(), HashSHA512(apiKey), types.TlsAuthType)
```

Comment:
should we make this function returns the created Bouncer to avoid selecting it just after ?